File: FastAPIUserService/routers/history.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from FastAPIUserService.config.db_config import get_db
from FastAPIUserService.crud.history import check_history, clearing_news_history, getting_news_history, toggling_view_news, add_view_news
from FastAPIUserService.schemas.history import HistoryCheckResponse
from FastAPIUserService.models.users import User
from FastAPIUserService.utils.auth import get_current_user
from FastAPIUserService.utils.success_response import success_response

router = APIRouter(prefix="/api/history", tags=["history"])


# 检查用户是否浏览了新闻
@router.get("/check")
async def check_viewed_history(
    news_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    is_viewed = await check_history(news_id, user.id, db)
    return success_response(message="检查历史浏览记录获取成功", data=HistoryCheckResponse(isView=is_viewed))


# 浏览新闻时添加到新闻历史浏览记录
@router.post("/toggle_view")
async def toggle_view_news(
    news_id: int,
    user: User =  Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    is_viewed = await toggling_view_news(user.id, news_id, db)
    return success_response(message="浏览新闻成功", data=HistoryCheckResponse(isView=is_viewed))


# 获取新闻历史浏览记录列表
@router.get("/history_list")
async def get_history_list(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    view_news_history_list = await getting_news_history(user.id, db)
    result = [
    {
        "id": n.id,
        "title": n.title,
        "image": n.image,
        "author": n.author,
        "publish_time": n.publish_time,
        "views": n.views,
        "categoryId": n.category_id
    }
    for n in view_news_history_list
    ]
    return success_response(message="获取收藏列表成功", data=result)



# 浏览新闻时添加浏览记录
@router.post("/add_view")
async def add_view_news_endpoint(
    news_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    is_viewed = await add_view_news(user.id, news_id, db)
    return success_response(message="浏览新闻成功", data=HistoryCheckResponse(isView=is_viewed))


# 清除浏览历史记录路由
@router.delete("/clear_history")
async def clear_news_history(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    count = await clearing_news_history(user.id, db)
    return success_response(message=f"清除了{count}历史浏览记录成功")

File: xwzx-news/src/api/history.js
/**
 * 历史浏览记录相关 API 接口
 */

// 检查是否浏览过某条新闻
export async function checkHistory(newsId) {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch(`/api/history/check?news_id=${newsId}`, {
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '查询失败')
  return data.data.isView
}

// 添加浏览记录（只添加，不会取消）
export async function addViewHistory(newsId) {
  const token = localStorage.getItem('user-token')
  if (!token) return // 未登录不记录
  const res = await fetch(`/api/history/add_view?news_id=${newsId}`, {
    method: 'POST',
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) {
    console.warn('记录浏览历史失败:', data.message)
  }
  return data
}

// 切换浏览记录（有则删，无则加 — 用于单条删除）
export async function toggleViewHistory(newsId) {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch(`/api/history/toggle_view?news_id=${newsId}`, {
    method: 'POST',
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '操作失败')
  return data.data.isView
}

// 获取浏览历史列表
export async function fetchHistoryList() {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch('/api/history/history_list', {
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '获取浏览历史失败')
  return data.data
}

// 清空浏览历史
export async function clearHistory() {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch('/api/history/clear_history', {
    method: 'DELETE',
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '清空失败')
  return data.message
}
